virtual_rainforest/core/grid.py

```diff
@@ -446,3 +452,132 @@ def populate_distances(
         """
 
         self._distances = squareform(pdist(self.centroids))
+
+    def map_xy_to_cell_id(
+        self,
+        x_coords: np.ndarray,
+        y_coords: np.ndarray,
+    ) -> list[list[Optional[int]]]:
+        """Map a set of coordinates onto grid cells.
+
+        This function loops over points defined by pairs of x and y coordinates and maps
+        the coordinates onto the cell_ids of the grid. The method also checks to see
+        that each point intersects one and only one of the cell polygons defined in the
+        grid. Points that intersect no cells fall outside the grid polygons and points
+        that intersect more than one cell fall ambiguously on cell borders.
+
+        Args:
+            x_coords: A numpy array of x coordinates of points that should occur within
+                grid cells.
+            y_coords: A similar and equal-length array providing y coordinates.
+
+        Returns:
+            A list of lists showing the cell ids that map onto each point. The list for
+            a given point can be empty - when the point falls in no cell - or of length
+            > 1 when a point falls on adjoining cell boundaries.
+        """
+
+        if (x_coords.ndim != 1) or (y_coords.ndim != 1):
+            raise ValueError("The x/y coordinate arrays are not 1 dimensional")
+
+        if x_coords.shape != y_coords.shape:
+            raise ValueError("The x/y coordinates are of unequal length")
+
+        # Get shapely points for the coordinates
+        xyp = [Point(x, y) for x, y in zip(x_coords, y_coords)]
+
+        # Map the Cell ID of each point - this is doing all pairs of points and cells,
+        # which is a computational choke point. Might be able to use a spatial search
+        # option, if this gets problematic, possibly including an STRTree in the grid
+        #    object https://shapely.readthedocs.io/en/latest/strtree.html
+
+        cell_map = [
+            [id for id, ply in zip(self.cell_id, self.polygons) if ply.intersects(pt)]
+            for pt in xyp
+        ]
+
+        return cell_map
+
+    def map_xy_to_cell_indexing(
+        self,
+        x_coords: np.ndarray,
+        y_coords: np.ndarray,
+        x_idx: Optional[np.ndarray],
+        y_idx: Optional[np.ndarray],
+        strict: Optional[bool] = True,
```

Comment:
Does `strict` need to be an `Optional` type? What's the difference in behaviour between when it's `None` vs `False`?